src/tools/info/proto_info.c

```diff
@@ -93,25 +93,211 @@ static void print_resource_usage(const resource_usage_t *usage_before,
+    ucs_status_t status;
+
+    request_param.op_attr_mask = UCP_OP_ATTR_FIELD_FLAGS;
+    request_param.flags        = flags;
+
+    status_ptr = ucp_ep_close_nbx(ep, &request_param);
+    status     = wait_completion(worker, status_ptr);
+    if (status != UCS_OK) {
+        printf("<Failed to destroy %s UCP endpoint>\n", ep_type);
+    }
+
+    return status;
+}
+
+static ucs_status_t
+create_listener(ucp_worker_h worker, ucp_listener_h *listener_p,
+                uint16_t *listen_port_p, void *accept_cb_arg)
+{
+    ucp_listener_h listener;
+    struct sockaddr_in listen_saddr;
+    ucp_listener_params_t listen_params;
+    ucp_listener_attr_t listen_attr;
     ucs_status_t status;
+
+    set_saddr("0.0.0.0", 0, &listen_saddr);
+
+    listen_params.field_mask         = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
+                                       UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER;
+    listen_params.sockaddr.addr      = (const struct sockaddr*)&listen_saddr;
+    listen_params.sockaddr.addrlen   = sizeof(listen_saddr);
+    listen_params.accept_handler.cb  = listener_accept_callback;
+    listen_params.accept_handler.arg = accept_cb_arg;
+
+    status = ucp_listener_create(worker, &listen_params, &listener);
+    if (status != UCS_OK) {
+        printf("<Failed to create UCP listener>\n");
+        goto out;
+    }
+
+    listen_attr.field_mask = UCP_LISTENER_ATTR_FIELD_SOCKADDR;
+
+    status = ucp_listener_query(listener, &listen_attr);
+    if (status != UCS_OK) {
+        printf("<Failed to query UCP listener>\n");
+        goto out_destroy_listener;
+    }
+
+    status = ucs_sockaddr_get_port((struct sockaddr*)&listen_attr.sockaddr,
+                                   listen_port_p);
+    if (status != UCS_OK) {
+        printf("<Failed to get port>\n");
+        goto out_destroy_listener;
+    }
+
+    *listener_p = listener;
+out:
+    return status;
+
+out_destroy_listener:
+    ucp_listener_destroy(listener);
+    goto out;
+}
+
+ucs_status_t
+print_ucp_ep_info(ucp_worker_h worker, const ucp_ep_params_t *base_ep_params,
+                  const char *ip_addr)
+{
+    ucp_listener_h listener    = NULL;
+    ucp_ep_h server_ep         = NULL;
+    ucp_address_t *worker_addr = NULL;
+    ucp_ep_params_t ep_params  = *base_ep_params;
+    ucs_status_t status, tmp_status;
     ucs_status_ptr_t status_ptr;
+    size_t worker_addr_length;
+    struct sockaddr_in connect_saddr;
+    uint16_t listen_port;
+    ucp_ep_h ep;
+    char ep_name[64];
+    ucp_request_param_t request_param;
+
+    if (ip_addr != NULL) {
+        status = create_listener(worker, &listener, &listen_port, &server_ep);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        ucs_strncpy_zero(ep_name, "client", sizeof(ep_name));
+
+        set_saddr(ip_addr, listen_port, &connect_saddr);
+
+        ep_params.field_mask      |= UCP_EP_PARAM_FIELD_FLAGS |
+                                     UCP_EP_PARAM_FIELD_SOCK_ADDR;
```

Comment:
maybe add another option to enable error handling? sock_addr usually is used together with err handling mode peer

Reply:
we already have `e` in features, it is set as part of `base_ep_params`
```
...
ucp_ep_params_t ep_params  = *base_ep_params;
...
```